text/0000-improved-ember-registry-apis.md

```diff
@@ -0,0 +1,556 @@
+
+```ts
+export interface LookupOptions {
+  singleton?: boolean;
+  instantiate?: boolean;
+  source?: string;
+  namespace?: string;
+}
+
+interface RegisterOptions {
+  singleton?: boolean;
+  instantiate?: boolean;
+}
+```
+
+#### `Factory` and `FactoryManager`
+
+Some APIs refer to *factories* and *factory managers*. The API for these classes is the same as it was previously, with the exception that the `fullName` and `normalizedNames` are deprecated:
+
+```ts
+interface FactoryClass {
+  positionalParams?: string | string[] | undefined | null;
+}
+
+interface Factory<T, C extends FactoryClass | object = FactoryClass> {
+  class?: C;
+  fullName?: string; // DEPRECATED
+  normalizedName?: string; // DEPRECATED
+  create(props?: { [prop: string]: any }): T;
+}
+
+interface FactoryManager<T = object> {
+  readonly class: Factory<T>;
+  readonly fullName?: string; // DEPRECATED
+  readonly normalizedName?: string; // DEPRECATED
+  create(props?: object): T;
+}
+```
+
+**TODO: nail down what here is public API during RFC discussion.**
+
+#### `Owner` API diff
+
+```diff
+ interface Owner {
+-  factoryFor(fullName: string, options: LookupOptions): FactoryManager;
++  factoryFor(identifier: Identifier, options: LookupOptions): FactoryManager;
+
+-  hasRegistration(fullName: string): boolean;
++  hasRegistration(identifier: Identifier): boolean;
+
+-  inject(factoryNameOrType: string, property: string, injectionName: string): void;
++  inject(factory: Identifier | FactoryTypeIdentifier, property: string, injection: Identifier): void;
+
+-  lookup(fullName: string, options?: LookupOptions): any;
++  lookup(identifier: Identifier, options?: LookupOptions): any;
+
+-  register(fullName: string, factory: any, options?: RegisterOptions): void;
++  register(identifier: Identifier, factory: any, options?: RegisterOptions): any;
+
+-  registerOptions(fullName: string, options: RegisterOptions): void;
++  registerOptions(identifier: Identifier, options: RegisterOptions): any;
+
+-  registerOptionsForType(fullName: string, options: RegisterOptions): void;
++  registerOptionsForType(identifier: FactoryTypeIdentifier, options: RegisterOptions): any;
+
+-  registeredOption(fullName: string, optionName: string): RegisterOptions;
++  registeredOption(identifier: Identifier, optionName: string): RegisterOptions;
+
+-  registeredOptions(fullName: string): RegisterOptions;
++  registeredOptions(identifier: Identifier): RegisterOptions;
+
+-  registeredOptionsForType(type: string): RegisterOptions;
++  registeredOptionsForType(type: FactoryTypeIdentifier): RegisterOptions;
+
+-  resolveRegistration(type: string): Factory;
++  resolveRegistration(identifier: Identifier): Factory;
+
+-  unregister(fullName: string): void;
++  unregister(identifier: Identifier): void;
+ }
+```
+
+### Codemod
+
+All existing microsyntax invocations can be straightforwardly migrated to the new syntax with a codemod.
```

Comment:
I think this is largely true, but not 100%. Specifically cases where the string is DRY'ed up and an imported variable. In those cases the codemod can't disambiguate between that argument being a string or a POJO.

I don't think its super common, but seems good to make it clear that the codemod can convert 100% of "static string" arguments to these APIs, and _some_ dynamism (but not all).

Reply:
Ah, excellent clarification. I'll update the text here to capture that more clearly.